distribution/xfer/download.go

```diff
@@ -0,0 +1,299 @@
+			topDownload.Transfer.Release(progressChan)
+			return initialRootFS, func() {}, ctx.Err()
+		case <-topDownload.Done():
+			break selectLoop
+		}
+	}
+
+	l, err := topDownload.result()
+	if err != nil {
+		topDownload.Transfer.Release(progressChan)
+		return initialRootFS, func() {}, err
+	}
+
+	for l != nil {
+		initialRootFS.DiffIDs = append([]layer.DiffID{l.DiffID()}, initialRootFS.DiffIDs...)
+		l = l.Parent()
+	}
+	return initialRootFS, func() { topDownload.Transfer.Release(progressChan) }, err
+}
+
+func (ldm *LayerDownloadManager) makeDownloadFunc(descriptor DownloadDescriptor, parentLayer layer.ChainID, parentDownload *downloadTransfer) DoFunc {
+	return func(progressChan chan<- Progress, start <-chan struct{}, inactive chan<- struct{}) Transfer {
+		d := &downloadTransfer{
+			Transfer:   NewTransfer(),
+			layerStore: ldm.layerStore,
+		}
+
+		go func() {
+			defer func() {
+				close(progressChan)
+			}()
+
+			select {
+			case <-start:
+			default:
+				progressChan <- downloadMessage(descriptor, "Waiting")
+				<-start
+			}
+
+			if parentDownload != nil {
+				// Did the parent download already fail or get
+				// cancelled?
+				select {
+				case <-parentDownload.Done():
+					_, err := parentDownload.result()
+					if err != nil {
+						d.err = err
+						return
+					}
+				default:
+				}
+			}
+
+			var (
+				downloadReader io.ReadCloser
+				size           int64
+				err            error
+				retries        int
+			)
+
+			for {
+				downloadReader, size, err = descriptor.Download(d.Transfer.Context(), progressChan)
+				if err == nil {
+					break
+				}
+				retries++
+				if _, isDNR := err.(DoNotRetry); isDNR || retries == maxDownloadAttempts {
+					d.err = err
+					return
+				}
+
+				delay := retries * 5
+				ticker := time.NewTicker(time.Second)
+				for {
+					progressChan <- downloadMessage(descriptor, fmt.Sprintf("Retrying in %d seconds", delay))
+					delay--
+					if delay == 0 {
+						ticker.Stop()
+						break
+					}
+					<-ticker.C
+				}
+			}
+
+			defer downloadReader.Close()
+
+			close(inactive)
+
+			if parentDownload != nil {
+				select {
+				case <-d.Transfer.Done():
+					d.err = errors.New("layer registration cancelled")
+					return
+				case <-parentDownload.Done():
+				}
+
+				l, err := parentDownload.result()
+				if err != nil {
+					d.err = err
+					return
+				}
+				parentLayer = l.ChainID()
+			}
+
+			reader := NewProgressReader(downloadReader, progressChan, size, descriptor.ID(), "Extracting")
+
+			inflatedLayerData, err := archive.DecompressStream(reader)
+			if err != nil {
+				d.err = fmt.Errorf("could not get decompression stream: %v", err)
+				return
+			}
+
+			registerDone := make(chan struct{})
+
+			go func() {
+				select {
+				case <-d.Transfer.Done():
+					inflatedLayerData.Close()
+				case <-registerDone:
+				}
+			}()
+
+			d.layer, err = d.layerStore.Register(inflatedLayerData, parentLayer)
+			close(registerDone)
+			if err != nil {
```

Comment:
Do you think it makes sense to retry in this case? There are two common reasons I can think of why `Register` would fail: disk full, or a corrupt tar stream. In either case, retrying the `Register` won't help, and neither will retrying the download, because the checksum was already verified (except in v1, but I want to avoid adding v1-specific behavior).